modules/proxy/httpsdialer.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package proxy
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/net/proxy"
+)
+
+// httpProxy is a HTTPS connect proxy.
+type httpProxy struct {
+	host     string
+	haveAuth bool
+	username string
+	password string
+	forward  proxy.Dialer
+}
+
+func newHTTPProxy(uri *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	s := new(httpProxy)
+	s.host = uri.Host
+	s.forward = forward
+	if uri.User != nil {
+		s.haveAuth = true
+		s.username = uri.User.Username()
+		s.password, _ = uri.User.Password()
+	}
+
+	return s, nil
+}
+
+// Dial  dials with http CONNECT method
+func (s *httpProxy) Dial(network, addr string) (c net.Conn, err error) {
+	// Dial and create the https client connection.
+	succeed := false
+	defer func() {
+		if !succeed && c != nil {
+			c.Close()
+		}
+	}()
+
+	c, err = s.forward.Dial("tcp", s.host)
+	if err != nil {
+		return nil, err
+	}
+
+	// HACK. http.ReadRequest also does this.
+	reqURL, err := url.Parse("http://" + addr)
+	if err != nil {
+		return nil, err
+	}
+	reqURL.Scheme = ""
+
+	req, err := http.NewRequest("CONNECT", reqURL.String(), nil)
```

Comment:
Seems we could use  [`httplib.NewRequest`](https://github.com/go-gitea/gitea/blob/6c501b149809a7015e58d9627e84bfd593702f95/modules/httplib/request.go#L36) instead, which also brings the benefit of setting a common user-agent.

Reply:
Sorry for late. As I check the `httplib.NewRequest`,  It aims to wrap a normal `HTTP request` and it uses to send a normal request internally :

```go
		TLSClientConfig: r.setting.TLSClientConfig,
		Proxy:           http.ProxyFromEnvironment,

```

while herein  the `proxy/httpsdialer.go` just make a `CONNECT` to the proxy only.